src/GafferScene/Instancer.cpp

```diff
@@ -269,6 +492,129 @@ class Instancer::EngineData : public Data
 			return result;
 		}
 
+		typedef std::vector< boost::unordered_set< IECore::MurmurHash > > PrototypeHashes;
+
+		// In order to compute the number of variations, we compute a unique hash for every context we use
+		// for evaluating prototypes.  So that we can track which sources are responsible for variations,
+		// we return a vector of hash sets, corresponding to the order of m_prototypeContextVariables,
+		// plus an extra entry at the end for the combined result of all variation sources
+		std::shared_ptr<PrototypeHashes> uniquePrototypeHashes() const
+		{
+			std::shared_ptr<PrototypeHashes> result( new PrototypeHashes( m_prototypeContextVariables.size() + 1 ) );
+
+			std::vector< IECore::MurmurHash> prototypeHashes;
+			prototypeHashes.reserve( m_prototypeIndexRemap.size() );
+			for( int indexRemap : m_prototypeIndexRemap )
+			{
+				IECore::MurmurHash h;
+				// We throw away rootStrings during initPrototypes, necessitating interating
+				// all the path components here - if perf was important, it would be better to
+				// store rootStrings
+				InternedStringVectorDataPtr path = m_roots[ indexRemap ];
+				h.append( &(path->readable())[0], path->readable().size() );
+				prototypeHashes.push_back( h );
+			}
+
+
+			size_t n = numPoints();
+			std::vector< IECore::MurmurHash > contextVariableHashes( m_prototypeContextVariables.size() );
+			for( unsigned int i = 0; i < n; i++ )
+			{
+				IECore::MurmurHash totalHash;
+				if( m_indices )
+				{
+					totalHash = prototypeHashes[ ((*m_indices)[i]) % m_numPrototypes ];
+				}
```

Comment:
Is there a problem here for `PrototypeMode::RootPerVertex`? In that case I think the primvar specifying the roots may not be indexed. That leads to `m_indices == nullptr`, but we would still need consider. `prototypeHashes[i]`. I don't think the same is true in the other modes. In those, null indices implies that it's always the first root that is used. I may be reading this wrong, but if not, would be good to add some tests to expose this.

Reply:
Note to self - I think you're right, still need to test and fix this.

Reply:
I don't think I got my head around the data flow well enough to figure out if there was a way of implementing the required logic centrally or if we needed to reproduce the logic here. Would be great to centralise if possible, but either way I think this probably warrants some test coverage.